compiler/src/iree/compiler/Codegen/Dialect/CPU/IR/IREECPUAttrs.cpp

```diff
@@ -0,0 +1,141 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Codegen/Dialect/CPU/IR/IREECPUDialect.h"
+#include "iree/compiler/Codegen/Dialect/CPU/IR/IREECPUTypes.h"
+#include "iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenTypes.h"
+#include "iree/compiler/Codegen/Dialect/Codegen/Utils/Utils.h"
+#include "iree/compiler/Dialect/Encoding/IR/EncodingTypes.h"
+#include "llvm/Support/Debug.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/IR/Attributes.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/Support/LLVM.h"
+
+#define DEBUG_TYPE "iree-cpu-attrs"
+#define DBGS() (llvm::dbgs() << "[" DEBUG_TYPE "]: ")
+#define LDBG(X) LLVM_DEBUG(DBGS() << X << "\n")
+
+#define GET_ATTRDEF_CLASSES
+#include "iree/compiler/Codegen/Dialect/CPU/IR/IREECPUAttrs.cpp.inc"
+
+namespace mlir::iree_compiler::IREE::CPU {
+
+using Codegen::MaterializeEncodingInfo;
+using Codegen::TileMxNxK;
+
+//===----------------------------------------------------------------------===//
+// iree_cpu.vmvx_encoding_layout
+//===----------------------------------------------------------------------===//
+
+// Enumerate tile sizes to choose from when no specific architecture is
+// targeted. For narrow-{M,N} cases, this only enumerates on narrow M. The
+// narrow-N cases are handled by transposition in chooseMatmulTile.
+static SmallVector<TileMxNxK>
+enumerateMatmulTilesVMVX(IREE::Encoding::EncodingAttr encoding,
+                         DictionaryAttr target) {
+  SmallVector<TileMxNxK> nonUkernelCandidates = {
+      TileMxNxK{8, 8, 4}, // Some vaguely reasonable tile shape.
+      TileMxNxK{4, 8, 4}, // Truncation of the above.
+      TileMxNxK{2, 8, 4}, // Truncation of the above.
+      TileMxNxK{1, 8, 4}, // Truncation of the above.
+  };
+  if (!target) {
+    return nonUkernelCandidates;
+  }
+  auto cDims = getEncodingContractionDims(encoding);
+  if (failed(cDims)) {
+    return nonUkernelCandidates;
+  }
+
+  auto ukernelAttr = target.getNamed("ukernels");
+  bool hasUkernelSupport = false;
+  if (ukernelAttr) {
+    auto strAttr = dyn_cast<StringAttr>(ukernelAttr->getValue());
+    if (strAttr && strAttr.getValue() == "all") {
+      hasUkernelSupport = true;
+    }
+  }
```

Comment:
I was wrong. The RISCV32 enables data-tiling only when ukernel is enabled. @lundong @rednoah91 Are we able to drop the check? I don't know if we have active users on this path or not. Otherwise, I'll make a call myself.

https://github.com/iree-org/iree/blob/4ee7d600c9ef024f85c7f6860abdebc1a964f7bc/compiler/src/iree/compiler/Codegen/Common/CPU/CPUMaterializeEncodings.cpp#L74-L86